File: C/14_Backtracking/BOJ_16198_에너지모으기.c
#include <stdio.h>
#define math_max(a, b) a > b ? a : b
int n, max = 0, arr[101], visited[101];
void dfs(int dpth, int sum) {
	if (dpth == n - 2) {
		max = math_max(max, sum);
		return;
	}
	for (int i = 1; i < n - 1; i++) {
		if (!visited[i]) {
			visited[i] = 1;
			int l = i, r = i;
			while (visited[l]) l--;
			while (visited[r]) r++;
			dfs(dpth + 1, sum + arr[l] * arr[r]);
			visited[i] = 0;
		}
	}
}
main() {
	scanf("%d", &n);
	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
	}
	dfs(0, 0);
	printf("%d\n", max);
}

File: JAVA/14_Backtracking/BOJ_16198_에너지모으기.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {

	static int n, max = 0;

	public static void dfs(int dpth, ArrayList<Integer> list, int sum) {
		if (dpth == n - 2) {
			max = Math.max(max, sum);
			return;
		}
		for (int i = 1; i < list.size() - 1; i++) {
			int energy = list.get(i - 1) * list.get(i + 1);
			int tmp = list.get(i);
			list.remove(i);
			dfs(dpth + 1, list, sum + energy);
			list.add(i, tmp);
		}
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt(br.readLine());
		ArrayList<Integer> list = new ArrayList<>();
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");
		for (int i = 0; i < n; i++) {
			list.add(Integer.parseInt(st.nextToken()));
		}
		dfs(0, list, 0);
		System.out.println(max);
	}
}